### PR TITLE
Log total number of tokens seen

### PR DESCRIPTION
**Pull request by runame**
I added logging of the total number of tokens that have been used for training up to the current `log` call. This is useful for plotting metrics like the loss against the number of tokens instead of steps.

Since `metrics_processor` is part of the `Trainer`'s state, this change should be compatible with checkpointing and resuming the count of `ntokens_seen`.

### CODE REVIEW

**Review by tianyu-l — COMMENTED** [review]
Change sounds OK to me, although I don't think it is true
> Since metrics_processor is part of the Trainer's state, this change should be compatible with checkpointing and resuming the count of ntokens_seen.

You have to explicitly put it in `state_dict` to be saved / loaded.
https://github.com/pytorch/torchtitan/blob/main/torchtitan/train.py#L574-L578

I'm not very sure if we want to put `self.metrics_processor.ntokens_seen` in the `Trainer` state dict, when other metrics are not put there.

**Comment by runame** [comment]
> You have to explicitly put it in state_dict to be saved / loaded.

You are right, I missed the `state_dict` definition, thanks!

> I'm not very sure if we want to put self.metrics_processor.ntokens_seen in the Trainer state dict, when other metrics are not put there.

The core difference is that this metric is accumulated across steps, so it is necessary to save it when a checkpoint is created to be able to resume logging of this metric correctly. This also the case for `step`.

Would you prefer if `ntokens_seen` was accumulated as an attribute of `Trainer`, like `step`? I think this might be cleaner in this case.

**Comment by tianyu-l** [comment]
> Would you prefer if ntokens_seen was accumulated as an attribute of Trainer, like step? I think this might be cleaner in this case.

Yeah let's do that!

File: torchtitan/train.py [hunk 1]
```diff
@@ -489,11 +491,13 @@ def train_step(
         else:
             global_avg_loss = global_max_loss = loss.detach().item()
 
+        self.ntokens_seen += self.metrics_processor.ntokens_since_last_log
```
Comment (tianyu-l) [inline torchtitan/train.py]: Let's not use the field from `metrics_processor`.
In `train.py` there is a line
```
self.metrics_processor.ntokens_since_last_log += labels.numel()
```
Let's reuse `labels.numel()`.

File: torchtitan/components/metrics.py [hunk 2]
```diff
@@ -359,6 +359,7 @@ def log(
         global_avg_loss: float,
         global_max_loss: float,
         grad_norm: float,
+        ntokens_seen: int,
```
Comment (tianyu-l) [inline torchtitan/components/metrics.py]: I might sound contradicting myself compared with your last PR on grad_norm:

Given that we don't print it to terminal (so nothing needs to be hardcoded), maybe let's just put it in `extra_metrics` instead of adding a new field.

File: torchtitan/train.py (hunk 1 above)
Comment (runame) [reply torchtitan/train.py]: Ok, I was debating this as well.

**Review by tianyu-l — APPROVED** [review]
Looks nice

**Comment by runame** [comment]
Thanks for the fast review!
